### PR TITLE
[hlslpp] Add new port

### PR DESCRIPTION
**Pull request by plekakis**
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by plekakis** [comment]
@microsoft-github-policy-service agree [company="default"]


Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by plekakis** [comment]
@microsoft-github-policy-service agree

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

File: ports/hlslpp/hlslpp-config.cmake [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# Set the version number
+set(hlslpp_VERSION 3.5)
+
+# Specify include directories
+set(hlslpp_INCLUDE_DIRS "${CMAKE_CURRENT_LIST_DIR}/../../include/hlslpp/include")
```
Comment (jimwang118) [inline ports/hlslpp/hlslpp-config.cmake:5]:
```suggestion
set(hlslpp_INCLUDE_DIRS "${CMAKE_CURRENT_LIST_DIR}/../../include/hlslpp/include")

```

File: ports/hlslpp/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+file(INSTALL "${SOURCE_PATH}/include" DESTINATION "${CURRENT_PACKAGES_DIR}/include/hlslpp/")
+
+# Copy and rename License -> copyright.
+file(INSTALL "${SOURCE_PATH}/LICENSE" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" RENAME copyright)
+
+# Copy the cmake config file, which is used to set the version and include directory vars.
+file(INSTALL "${CMAKE_CURRENT_LIST_DIR}/${PORT}-config.cmake" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
+
+# Copy and show the usage.
+configure_file("${CMAKE_CURRENT_LIST_DIR}/usage" "${CURRENT_PACKAGES_DIR}/share/${PORT}/usage" COPYONLY)
```
Comment (jimwang118) [inline ports/hlslpp/portfile.cmake:21]:
```suggestion
configure_file("${CMAKE_CURRENT_LIST_DIR}/usage" "${CURRENT_PACKAGES_DIR}/share/${PORT}/usage" COPYONLY)

```

File: ports/hlslpp/usage [hunk 3]
```diff
@@ -0,0 +1,4 @@
+hlslpp provides header-only CMake targets:
+
+find_package(hlslpp CONFIG REQUIRED)
+target_include_directories(main PUBLIC ${hlslpp_INCLUDE_DIRS})
```
Comment (jimwang118) [inline ports/hlslpp/usage:4]:
```suggestion
target_include_directories(main PUBLIC ${hlslpp_INCLUDE_DIRS})

```

**Review by jimwang118 — CHANGES_REQUESTED** [review]

**Review by jimwang118 — APPROVED** [review]

File: ports/hlslpp/vcpkg.json [hunk 4]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "hlslpp",
+  "version": "3.5",
+  "port-version": 2,
```
Comment (jimwang118) [inline ports/hlslpp/vcpkg.json:4]:
```suggestion
```
The newly added port does not require this keyword.

File: versions/h-/hlslpp.json [hunk 5]
```diff
@@ -0,0 +1,14 @@
+{
+  "versions": [
+    {
+      "git-tree": "e3b4749575931138c4affb82aa00fedf63f78287",
+      "version": "3.5",
+      "port-version": 2
+    },
```
Comment (jimwang118) [inline versions/h-/hlslpp.json:7]:
```suggestion
```

File: ports/hlslpp/portfile.cmake [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    REF "${VERSION}"
+    SHA512 79b1464eb6441ec720b9952e80d7fbdb67852f7770fbe6ec2c67736627e62292f9d23dfcc9b1d548b6c33b75a7b2e911fa757fe087d7360bc7e72867d7f2c7a8
+    HEAD_REF master
+)
+
+# This is a header only library, copy them over.
+file(INSTALL "${SOURCE_PATH}/include" DESTINATION "${CURRENT_PACKAGES_DIR}/include/hlslpp/")
+
+# Copy and rename License -> copyright.
+file(INSTALL "${SOURCE_PATH}/LICENSE" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" RENAME copyright)
```
Comment (jimwang118) [inline ports/hlslpp/portfile.cmake:15]:
```suggestion
vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/LICENSE")
```

**Review by jimwang118 — CHANGES_REQUESTED** [review]

**Comment by jimwang118** [comment]
Note: I will be converting your PR to draft status. The above suggested changes are only recommendations. If you are willing to adopt them, please click "ready for review" after making the modifications. If you do not wish to make any changes, please click "ready for review" directly. That way, I can be aware that you've responded since you can't modify the tags.

File: ports/hlslpp/usage [hunk 7]
```diff
@@ -0,0 +1,4 @@
+hlslpp provides header-only CMake targets:
```
Comment (dg0yt) [inline ports/hlslpp/usage:1]: It doesn't provide targets.
```suggestion
hlslpp provides a CMake package:
```

File: ports/hlslpp/usage [hunk 8]
```diff
@@ -0,0 +1,4 @@
+hlslpp provides header-only CMake targets:
+
+find_package(hlslpp CONFIG REQUIRED)
+target_include_directories(main PUBLIC ${hlslpp_INCLUDE_DIRS})
```
Comment (dg0yt) [inline ports/hlslpp/usage:4]:
```suggestion
  find_package(hlslpp CONFIG REQUIRED)
  target_include_directories(main PUBLIC ${hlslpp_INCLUDE_DIRS})
```
As applied by vpckg tool heuristics.

File: ports/hlslpp/hlslpp-config.cmake [hunk 9]
```diff
@@ -0,0 +1,5 @@
+# Set the version number
+set(hlslpp_VERSION 3.5)
```
Comment (dg0yt) [inline ports/hlslpp/hlslpp-config.cmake:2]:
```suggestion
set(hlslpp_VERSION "@VERSION@")
```
and install with `configure_file ... @ONLY`. So it doesn't need updating again.

File: ports/hlslpp/hlslpp-config.cmake [hunk 10]
```diff
@@ -0,0 +1,5 @@
+# Set the version number
+set(hlslpp_VERSION 3.5)
+
+# Specify include directories
```
Comment (dg0yt) [inline ports/hlslpp/hlslpp-config.cmake:4]:
```suggestion
```

File: ports/hlslpp/hlslpp-config.cmake [hunk 11]
```diff
@@ -0,0 +1,5 @@
+# Set the version number
+set(hlslpp_VERSION 3.5)
+
+# Specify include directories
+set(hlslpp_INCLUDE_DIRS "${CMAKE_CURRENT_LIST_DIR}/../../include/hlslpp/include")
```
Comment (dg0yt) [inline ports/hlslpp/hlslpp-config.cmake:5]: Wouldn't it be nice to place includes in `${CMAKE_CURRENT_LIST_DIR}/../../include/hlslpp`?

File: ports/hlslpp/portfile.cmake [hunk 12]
```diff
@@ -0,0 +1,21 @@
+# vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
+
```
Comment (dg0yt) [inline ports/hlslpp/portfile.cmake:2]:
```suggestion
# vcpkg_check_linkage(ONLY_STATIC_LIBRARY)

```

File: ports/hlslpp/portfile.cmake [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO redorav/hlslpp
+    REF "${VERSION}"
+    SHA512 79b1464eb6441ec720b9952e80d7fbdb67852f7770fbe6ec2c67736627e62292f9d23dfcc9b1d548b6c33b75a7b2e911fa757fe087d7360bc7e72867d7f2c7a8
+    HEAD_REF master
+)
+
+# This is a header only library, copy them over.
+file(INSTALL "${SOURCE_PATH}/include" DESTINATION "${CURRENT_PACKAGES_DIR}/include/hlslpp/")
```
Comment (dg0yt) [inline ports/hlslpp/portfile.cmake:12]: Maybe
```suggestion
file(INSTALL "${SOURCE_PATH}/include" DESTINATION "${CURRENT_PACKAGES_DIR}/include/")
file(RENAME "${CURRENT_PACKAGES_DIR}/include/include" "${CURRENT_PACKAGES_DIR}/include/hlslpp")
```

**Review by dg0yt — CHANGES_REQUESTED** [review]
Most comments don't add value to the command that follow. Just remove them.

The CMake config package is not provided by upstream. The package-name should be prefixed with `unofficial-`.  
https://learn.microsoft.com/en-us/vcpkg/contributing/maintainer-guide#add-cmake-exports-in-an-unofficial--namespace  
And even for a header-only package, it might make sense to provide an imported target instead of variables.

**Comment by plekakis** [comment]
Apologies for all the above, this is my first vcpkg contribution and I am learning still.
I'm talking to the author of hlslpp and doing this in his place instead, so I would still like this to be an official port - bear with me and will sort it all out properly.

**Review by jimwang118 — DISMISSED** [review]

**Comment by jimwang118** [comment]
Usage test pass with following triplet:

```
x64-windows
```

File: ports/hlslpp/hlslpp-config.cmake [hunk 14]
```diff
@@ -0,0 +1,4 @@
+add_library(hlslpp INTERFACE)
+add_library(hlslpp::hlslpp ALIAS hlslpp)
+
+target_include_directories(hlslpp INTERFACE "${CMAKE_CURRENT_LIST_DIR}/../../include/hlslpp")
```
Comment (dg0yt) [inline ports/hlslpp/hlslpp-config.cmake:4]: The missing piece was the `IMPORTED` keyword.
And as long as it isn't provided by upstream, it package name and target name shall be prefixed as `unofficial`.
```suggestion
add_library(unofficial::hlslpp::hlslpp INTERACE IMPORTED)
target_include_directories(unofficial::hlslpp::hlslpp INTERFACE "${CMAKE_CURRENT_LIST_DIR}/../../include/hlslpp")
```
installed as `share/unofficial-hlslpp/unofficial-hlslpp-config.cmake`.

File: ports/hlslpp/usage [hunk 15]
```diff
@@ -0,0 +1,4 @@
+hlslpp provides a CMake package:
+
+  find_package(hlslpp CONFIG REQUIRED)
+  target_link_libraries(main PUBLIC hlslpp::hlslpp)
```
Comment (dg0yt) [inline ports/hlslpp/usage:4]:
```suggestion
  find_package(unofficial-hlslpp CONFIG REQUIRED)
  target_link_libraries(main PUBLIC unofficial::hlslpp::hlslpp)
```

**Pull request merged**